include/ginkgo/core/base/math.hpp

```diff
@@ -141,6 +162,98 @@ struct is_complex_impl<std::complex<T>>
     : public std::integral_constant<bool, true> {};
 
 
+template <typename T>
+struct is_complex_or_scalar_impl : std::is_scalar<T> {};
+
+template <typename T>
+struct is_complex_or_scalar_impl<std::complex<T>> : std::is_scalar<T> {};
+
+
+/**
+ * template_convertor is converting the template parameters of a class by
+ * convertor<type>.
+ *
+ * @tparam  convertor<type> which convert one type to another type
+ * @tparam  T  type
+ */
+template <template <typename> class convertor, typename T>
+struct template_convertor {};
+
+/**
+ * template_convertor is converting the template parameters of a class by
+ * convertor<type>. Converting class<T1, T2, ...> to class<convertor<T1>,
+ * convertor<T2>, convertor<...>>.
+ *
+ * @tparam  convertor<type> which convert one type to another type
+ * @tparam  template <...> T  class template base
+ * @tparam  ...Rest  the template parameter of T
+ */
+template <template <typename> class convertor, template <typename...> class T,
+          typename... Rest>
+struct template_convertor<convertor, T<Rest...>> {
+    using type = T<typename convertor<Rest>::type...>;
+};
+
+
+template <typename T, typename = void>
+struct remove_complex_s {};
+
+/**
+ * Obtains a real counterpart of a std::complex type, and leaves the type
+ * unchanged if it is not a complex type for complex/scalar type.
+ *
+ * @tparam T  complex or scalar type
+ */
+template <typename T>
+struct remove_complex_s<
+    T, xstd::void_t<std::enable_if_t<is_complex_or_scalar_impl<T>::value>>> {
```

Comment:
nit: you don't need the `xstd::void_t` here since `enable_if_t` already translates to the type `void` by default:
```suggestion
    T, std::enable_if_t<is_complex_or_scalar_impl<T>::value>> {
```